In [1]:
import pandas as pd 
import numpy as np
import os
from pandas import DataFrame, Series
import tabulate

PATH = os.getcwd()
FILE = "../output/FE_GBPUSD_15mins_1yr_End_20260311.csv"
df = pd.read_csv(FILE)
df["Date"] = pd.DatetimeIndex(df["Date"], tz="US/Eastern")
df.set_index("Date", inplace=True)
df


,Open,High,Low,Close,Idx,Body,Range,UWick,LWick,Close_%High,...,Bear_Engulf,Dark_Cloud,Piercing,Bull_BBR,Bear_BBR,ILR,IHR,S_R,BBU_BO,BBL_BO
Date,,,,,,,,,,,,,,,,,,,,,
2025-03-11 17:15:00-04:00,1.295630,1.295685,1.295180,1.295335,0,0.000295,0.000505,0.000055,0.000155,0.693069,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2025-03-11 17:30:00-04:00,1.295335,1.295360,1.294550,1.294780,1,0.000555,0.000810,0.000025,0.000230,0.716049,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2025-03-11 17:45:00-04:00,1.294780,1.294800,1.294590,1.294655,2,0.000125,0.000210,0.000020,0.000065,0.690476,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2025-03-11 18:00:00-04:00,1.294655,1.295075,1.294470,1.294995,3,0.000340,0.000605,0.000080,0.000185,0.132231,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2025-03-11 18:15:00-04:00,1.294995,1.295000,1.294595,1.294665,4,0.000330,0.000405,0.000005,0.000070,0.827160,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2026-03-11 15:45:00-04:00,1.341870,1.342230,1.341675,1.341855,24575,0.000015,0.000555,0.000360,0.000180,0.675676,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2026-03-11 16:00:00-04:00,1.341855,1.341855,1.341475,1.341765,24576,0.000090,0.000380,0.000000,0.000290,0.236842,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2026-03-11 16:15:00-04:00,1.341765,1.341770,1.341560,1.341580,24577,0.000185,0.000210,0.000005,0.000020,0.904762,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


Get al columns

In [2]:
df.columns

Index(['Open', 'High', 'Low', 'Close', 'Idx', 'Body', 'Range', 'UWick',
       'LWick', 'Close_%High', 'Open_%High', 'Iday_Idx', 'Iday_High',
       'Iday_Low', 'Iday_Range', 'Close_%DHigh', 'Open_%DHigh', 'Yday_High',
       'Yday_Low', 'Day_Idx', 'ADR', 'ATR', 'SMA4', 'SMA16', 'SMA32',
       'BB_Upper_16_2', 'BB_Lower_16_2', 'RSI', 'Sig_High', 'Sig_Low',
       'SMA_Trend', 'SMA16_Slope', 'SMA32_Slope', 'Hammer', 'Shooting_Star',
       'Bull_Engulf', 'Bear_Engulf', 'Dark_Cloud', 'Piercing', 'Bull_BBR',
       'Bear_BBR', 'ILR', 'IHR', 'S_R', 'BBU_BO', 'BBL_BO'],
      dtype='object')

Get all rows that have `Bear_BBR` as True

In [73]:
def get_yra_std(row: Series, df: DataFrame, n: int):
    std = df["YRA_Diff"].at_time("17:15").iloc[row["Day_Idx"]-n+1:row["Day_Idx"]+1].std()
    return std

pd.options.display.max_rows = 100
cols = ['Idx','Day_Idx', 'Open', 'High', 'Low', 'Close', 'Body', 'Range', 'ATR',
        'Close_%High', 'Open_%High', 'Iday_High', 'Iday_Low', 
        'Iday_Range', 'ADR', 'Close_%DHigh', 'Open_%DHigh', 
        'Yday_High', 'Yday_Low', 'Yday_Range', 'YRA_Diff', 'YRA_STD', 'SMA4', 'SMA16', 'SMA32',
       'BB_Upper_16_2', 'BB_Lower_16_2', 'RSI', 'Sig_High', 'Sig_Low',
       'SMA_Trend', 'SMA16_Slope', 'SMA32_Slope', 'Bear_BBR', 'Shooting_Star',
       'Bear_Engulf', 'Dark_Cloud', 'IHR', 'S_R', 'BBU_BO', 
       'ILR', 'Bull_BBR', 'Hammer', 'Bull_Engulf', 'BBL_BO']
pd.options.display.max_columns = len(cols)
df["Yday_Range"] = df.eval("Yday_High - Yday_Low")
df["YRA_Diff"] = df.apply(lambda x: x["Yday_Range"] - x["ADR"], axis=1)
df["YRA_STD"] = df.apply(get_yra_std, axis=1, args=[df, 30])



In [86]:
bbu_reversal = df[cols]
bbu_reversal.query("Bear_BBR == True")

,Idx,Day_Idx,Open,High,Low,Close,Body,Range,ATR,Close_%High,Open_%High,Iday_High,Iday_Low,Iday_Range,ADR,Close_%DHigh,Open_%DHigh,Yday_High,Yday_Low,Yday_Range,YRA_Diff,YRA_STD,SMA4,SMA16,SMA32,BB_Upper_16_2,BB_Lower_16_2,RSI,Sig_High,Sig_Low,SMA_Trend,SMA16_Slope,SMA32_Slope,Bear_BBR,Shooting_Star,Bear_Engulf,Dark_Cloud,IHR,S_R,BBU_BO,ILR,Bull_BBR,Hammer,Bull_Engulf,BBL_BO
Date,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
2025-03-12 05:00:00-04:00,47,0,1.294210,1.294530,1.293750,1.293895,0.000315,0.000780,0.000939,0.814103,0.410256,1.295685,1.291365,0.004320,NaN,0.414352,0.341435,NaN,NaN,NaN,NaN,NaN,1.293565,1.292857,1.292971,1.294028,1.291687,57.242400,1.295685,1.291365,1,35.509439,6.123933,True,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2025-03-12 07:30:00-04:00,57,0,1.295655,1.296145,1.295100,1.295430,0.000225,0.001045,0.000943,0.684211,0.468900,1.296145,1.291365,0.004780,NaN,0.149582,0.102510,NaN,NaN,NaN,NaN,NaN,1.294770,1.293912,1.293191,1.295513,1.292310,63.957243,1.295685,1.291365,2,50.534430,28.026001,True,NaN,NaN,NaN,True,1.0,NaN,NaN,NaN,NaN,NaN,NaN
2025-03-12 12:15:00-04:00,76,0,1.298070,1.298265,1.297430,1.297660,0.000410,0.000835,0.001290,0.724551,0.233533,1.299000,1.291365,0.007635,NaN,0.175508,0.121807,NaN,NaN,NaN,NaN,NaN,1.297094,1.295018,1.294655,1.297964,1.292071,65.991526,1.299000,1.291365,2,48.524029,46.140500,True,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2025-03-12 22:15:00-04:00,115,1,1.297265,1.297350,1.296885,1.297010,0.000255,0.000465,0.000436,0.731183,0.182796,1.297350,1.296075,0.001275,NaN,0.266667,0.066667,1.299000,1.291365,0.007635,NaN,NaN,1.296846,1.296547,1.296601,1.297143,1.295951,57.141452,NaN,NaN,1,13.473158,1.909152,True,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2025-03-13 04:30:00-04:00,140,1,1.296485,1.296485,1.295535,1.295910,0.000575,0.000950,0.000938,0.605263,0.000000,1.297350,1.294155,0.003195,NaN,0.450704,0.270736,1.299000,1.291365,0.007635,NaN,NaN,1.295755,1.295502,1.295990,1.296433,1.294571,50.937783,1.297350,1.296075,0,4.051684,-0.029842,True,NaN,NaN,NaN,NaN,4.0,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2026-03-06 16:15:00-05:00,24292,252,1.340140,1.340270,1.339910,1.339925,0.000215,0.000360,0.001030,0.958333,0.361111,1.340980,1.331155,0.009825,0.010079,0.107379,0.085496,1.338710,1.329750,0.008960,-0.001119,0.003966,1.339426,1.339091,1.337920,1.340132,1.338050,58.904509,1.340980,1.331155,2,19.978314,46.297646,True,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2026-03-09 02:00:00-04:00,24330,253,1.333365,1.334075,1.332485,1.332730,0.000635,0.001590,0.001371,0.845912,0.446541,1.335175,1.328310,0.006865,0.009884,0.356154,0.263656,1.341630,1.331155,0.010475,0.000591,0.003713,1.332796,1.330734,1.330619,1.333298,1.328170,55.186083,1.335175,1.328310,2,59.302734,0.835504,True,NaN,NaN,True,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2026-03-09 09:30:00-04:00,24360,253,1.337745,1.338270,1.336690,1.337575,0.000170,0.001580,0.001383,0.439873,0.332278,1.338270,1.328310,0.009960,0.009884,0.069779,0.052711,1.341630,1.331155,0.010475,0.000591,0.003713,1.336907,1.335188,1.333955,1.337667,1.332710,66.050386,1.336365,1.328310,2,56.383248,46.819677,True,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


# Bear_BBR Signal Analysis
    

## Bearish BBR Metrics Function
Return a new dataframe with all occrrences of Bear_BBR signal and get additional metrics such as:
- Entry Price
- Max Gain / Loss
- Gain/Loss to:
    - SMA16,
    - SMA32,
    - BB_Lower,
    - SigLow,
    - First_Bullish_Signal (Bull_0)
    - Bull_Engulf
    - Intraday_Low_Reversal (ILR)
- Whether S/R was tested or broken
- Whether other bearish pattern occurred at the same time

In [80]:

def bearish_bbr_metrics(df: DataFrame, bear_bbr_signal_idx: list):
    """Return a new dataframe with signal metrics"""

    signal = []
    for signal_idx in bear_bbr_signal_idx:
        row = df.iloc[signal_idx]
        s = {}
        START = row.name
        TD = pd.Timedelta(minutes=15)
        s["Date"] = START
        s["Entry_Price"] = row["Close"] # signal price
        
        day = row["Day_Idx"]
        window = df.query(f"Day_Idx == {day}") # only get df data from the same day as signal
        before_signal_window = window[:START]
        signal_window = window[START:]
        Max_High = signal_window[START+TD:]["High"].max() # Max high from signal until EOD
        Min_Low = signal_window[START+TD:]["Low"].min() # Min low from signal until EOD
        s["D_Max_Up"] = s["Entry_Price"] - Max_High # Max price gain from signal until EOD
        s["D_Max_Down"] = s["Entry_Price"] - Min_Low # Max price loss from signal until EOD
        D_Close = signal_window.iloc[-1]["Close"] # Closing price at EOD
        s["D_Gain"] = s["Entry_Price"] - D_Close # Gain at EOD from signal price


        if signal_window.query(f"Close > {row["High"]}").empty is False:
            # If price closed above signal price within the day, 
            # get the dataframe between start and failure time
            EXIT = signal_window[START:].query(f"Close > {row["High"]}").iloc[0].name
            entry_window = signal_window[START:EXIT]
        else:
            # If price didn't close above signal price within the day
            # get the dataframe from signal time until EOD
            entry_window = signal_window

        # Did the price go higher than the signal candle high
        if entry_window[START+TD:].query(f"High >= {row["High"]}").empty is False:
            s["Stop_Pips"] = s["Entry_Price"] - row["High"]
        else:
            s["Stop_Pips"] = 0
        
        s["Max_Up"] = s["Entry_Price"] - entry_window[START+TD:]["Close"].max()
        s["Max_Down"] = s["Entry_Price"] - entry_window[START+TD:]["Low"].min()
        

        # If the price fell to the SMA16 after the signal
        # calcuate the gain
        below_sma16 = entry_window.query("Low < SMA16")
        if below_sma16.empty is False:
            s["To_SMA16"] = s["Entry_Price"] - below_sma16.iloc[0]["SMA16"] 
        else: 
            s["To_SMA16"] = 0

        # If the price fell to the SMA32 after the signal
        # calcuate the gain
        below_sma32 = entry_window.query("Low < SMA32")
        if below_sma32.empty is False:
            s["To_SMA32"] = s["Entry_Price"] - below_sma32.iloc[0]["SMA32"]
        else:
            s["To_SMA32"] = 0

        # If the price fell to the BB_Lower_16_2 after the signal
        # calcuate the gain
        below_bbl = entry_window.query("Low < BB_Lower_16_2")
        if below_bbl.empty is False:
            s["To_BBL_16_2"] = s["Entry_Price"] - below_bbl.iloc[0]["BB_Lower_16_2"]
        else:
            s["To_BBL_16_2"] = 0

        # If the price fell to the Sig_Low after the signal
        # calcuate the gain
        below_sig_low = entry_window.query("Low < Sig_Low")
        if below_sig_low.empty is False:
            s["To_Sig_Low"] = s["Entry_Price"] - below_sig_low.iloc[0]["Sig_Low"]
        else:
            s["To_Sig_Low"] = 0

        below_sr = entry_window.query("S_R > 3")["Sig_Low"].unique()
        for i in range(len(below_sr)):
            s[f"To_S_{i}"] = s["Entry_Price"] - below_sr[i]


        # Check if bullish signals occured during winddow
        NULL_Timestamp = pd.Timestamp(0, tz="US/Eastern")
        to_bull_bbr = entry_window.query("Bull_BBR == True")
        if to_bull_bbr.empty is False:
            s["To_Bull_BBR"] = s["Entry_Price"] - to_bull_bbr.iloc[0]["Close"]
            to_bull_bbr_time_stamp = to_bull_bbr.iloc[0].name
        else:
            s["To_Bull_BBR"] = 0
            to_bull_bbr_time_stamp = NULL_Timestamp

        to_bull_engulf = entry_window.query("Bull_Engulf == True")
        if to_bull_engulf.empty is False:
            s["To_Bull_Engulf"] = s["Entry_Price"] - to_bull_engulf.iloc[0]["Close"]
            to_bull_engulf_time_stamp = to_bull_engulf.iloc[0].name
        else:
            s["To_Bull_Engulf"] = 0
            to_bull_engulf_time_stamp = NULL_Timestamp

        to_ilr = entry_window.query("ILR == True")
        if to_ilr.empty is False:
            s["To_ILR"] = s["Entry_Price"] - to_ilr.iloc[0]["Close"]
            to_ilr_time_stamp = to_ilr.iloc[0].name
        else:
            s["To_ILR"] = 0
            to_ilr_time_stamp = NULL_Timestamp

        first_bullish_signal = min([to_bull_bbr_time_stamp, to_bull_engulf_time_stamp, to_ilr_time_stamp])
        if first_bullish_signal > NULL_Timestamp:
            s["To_Bull_0"] = s["Entry_Price"] - entry_window.loc[first_bullish_signal]["Close"]
        else:
            s["To_Bull_0"] = 0


        # append other useful df metrics to new df
        s["Range"] = row["Range"]
        s["ATR"] = row["ATR"]
        s["Close_%High"] = row["Close_%High"]
        s["SMA_Trend"] = row["SMA_Trend"]
        s["SMA16_Slope"] = row["SMA16_Slope"]
        s["SMA32_Slope"] = row["SMA32_Slope"]
        s["RSI"] = row["RSI"]
        s["IHR"] = row["IHR"]
        s["Dark_Cloud"] = row["Dark_Cloud"]
        s["Shooting_Star"] = row["Shooting_Star"]
        s["Bear_Engulf"] = row["Bear_Engulf"]
        s["Iday_Range"] = row["Iday_Range"]
        s["ADR"] = row["ADR"]
        s["Close_%DHigh"] = row["Close_%DHigh"]
        s["SMA16"] = row["SMA16"]
        s["SMA32"] = row["SMA32"]
        s["S_R"] = row["S_R"]
        s["YRA_Diff"] = row["YRA_Diff"]
        s["YRA_STD"] = row["YRA_STD"]

        signal.append(s)
    sdf = pd.DataFrame(signal)
    return sdf


### Get Bear_BBR Metrics

In [81]:
signal_idx_list = bbu_reversal.query("Bear_BBR == True")["Idx"].to_list()
bearish_bbr_df = bearish_bbr_metrics(bbu_reversal, signal_idx_list)
bearish_bbr_df

,Date,Entry_Price,D_Max_Up,D_Max_Down,D_Gain,Stop_Pips,Max_Up,Max_Down,To_SMA16,To_SMA32,To_BBL_16_2,To_Sig_Low,To_Bull_BBR,To_Bull_Engulf,To_ILR,To_Bull_0,Range,ATR,Close_%High,SMA_Trend,SMA16_Slope,SMA32_Slope,RSI,IHR,Dark_Cloud,Shooting_Star,Bear_Engulf,Iday_Range,ADR,Close_%DHigh,SMA16,SMA32,S_R,YRA_Diff,YRA_STD,To_S_0,To_S_1,To_S_2
0,2025-03-12 05:00:00-04:00,1.293895,-0.005105,0.001435,-0.002435,-0.000635,-0.000740,0.000925,0.000602,0.000906,0.000000,0.000000,0.000000,-0.000305,0.000000,0.000000,0.000780,0.000939,0.814103,1,35.509439,6.123933,57.242400,NaN,NaN,NaN,NaN,0.004320,NaN,0.414352,1.292857,1.292971,NaN,NaN,NaN,NaN,NaN,NaN
1,2025-03-12 07:30:00-04:00,1.295430,-0.003570,0.002970,-0.000900,-0.000715,-0.001390,0.002970,0.001065,0.001654,0.002231,0.000000,0.000000,0.000000,0.000000,0.000000,0.001045,0.000943,0.684211,2,50.534430,28.026001,63.957243,True,NaN,NaN,NaN,0.004780,NaN,0.149582,1.293912,1.293191,1.0,NaN,NaN,NaN,NaN,NaN
2,2025-03-12 12:15:00-04:00,1.297660,-0.000450,0.001675,0.001330,0.000000,-0.000205,0.001675,0.000862,0.001504,0.001517,0.000000,0.000000,0.000000,0.000000,0.000000,0.000835,0.001290,0.724551,2,48.524029,46.140500,65.991526,NaN,NaN,NaN,NaN,0.007635,NaN,0.175508,1.295018,1.294655,NaN,NaN,NaN,NaN,NaN,NaN
3,2025-03-12 22:15:00-04:00,1.297010,-0.000120,0.004910,0.001915,0.000000,-0.000010,0.004910,0.000418,0.000410,0.001155,0.000935,0.001905,0.001055,0.000525,0.001905,0.000465,0.000436,0.731183,1,13.473158,1.909152,57.141452,NaN,NaN,NaN,NaN,0.001275,NaN,0.266667,1.296547,1.296601,NaN,NaN,NaN,0.000935,0.002855,NaN
4,2025-03-13 04:30:00-04:00,1.295910,-0.000980,0.003810,0.000815,-0.000575,-0.000045,0.003810,0.000498,-0.000080,0.001447,-0.000165,0.001785,-0.000045,0.000000,0.000000,0.000950,0.000938,0.605263,0,4.051684,-0.029842,50.937783,NaN,NaN,NaN,NaN,0.003195,NaN,0.450704,1.295502,1.295990,4.0,NaN,NaN,-0.000165,0.001755,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
354,2026-03-06 16:15:00-05:00,1.339925,-0.001705,0.000855,-0.000155,-0.000345,-0.000155,0.000855,0.000737,0.000000,0.000000,0.000000,0.000000,-0.000155,0.000000,0.000000,0.000360,0.001030,0.958333,2,19.978314,46.297646,58.904509,NaN,NaN,NaN,NaN,0.009825,0.010079,0.107379,1.339091,1.337920,NaN,-0.001119,0.003966,NaN,NaN,NaN
355,2026-03-09 02:00:00-04:00,1.332730,-0.011985,0.002475,-0.010660,-0.001345,-0.001735,0.002110,0.001554,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.001590,0.001371,0.845912,2,59.302734,0.835504,55.186083,NaN,True,NaN,NaN,0.006865,0.009884,0.356154,1.330734,1.330619,NaN,0.000591,0.003713,NaN,NaN,NaN
356,2026-03-09 09:30:00-04:00,1.337575,-0.007140,0.002990,-0.005815,-0.000695,-0.002150,0.002990,0.002020,0.000000,0.000000,0.000000,0.000000,0.000460,0.000000,0.000000,0.001580,0.001383,0.439873,2,56.383248,46.819677,66.050386,NaN,NaN,NaN,NaN,0.009960,0.009884,0.069779,1.335188,1.333955,NaN,0.000591,0.003713,0.009265,NaN,NaN
357,2026-03-09 11:15:00-04:00,1.339030,-0.005685,0.000830,-0.004360,-0.000750,-0.001385,0.000680,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.001240,0.001635,0.604839,2,60.917219,56.519957,65.245502,NaN,NaN,NaN,NaN,0.011760,0.009884,0.088435,1.336594,1.335089,NaN,0.000591,0.003713,NaN,NaN,NaN


## Signal Stats

### Initial Signal Stats

In [724]:
# get the total number of bear_bbr signals
total_signals_count = bearish_bbr_df[bearish_bbr_df.columns[0]].count()

day_closed_lower = bearish_bbr_df.query("D_Gain > 0")
day_closed_higher = bearish_bbr_df.query("D_Gain <= 0")
day_closed_lower_count = day_closed_lower["D_Gain"].count()
day_closed_higher_count = day_closed_higher["D_Gain"].count()
day_closed_lower_pct = round(day_closed_lower_count / total_signals_count * 100,2)
day_closed_lower_pips = bearish_bbr_df["D_Gain"].sum().round(6)

# filter the dataframe for occurences where after the signal
# the price moved lower more than it moved higher
d_max_down_greater = bearish_bbr_df.query("D_Max_Up.abs() < D_Max_Down")
# get the count of occurences where price dropped more then it rose after signal
d_max_down_greater_count = d_max_down_greater[d_max_down_greater.columns[0]].count()

# percentage of where price fell further than the max it went against the entry price
d_max_down_greater_pct = round(d_max_down_greater_count / total_signals_count * 100, 2)

max_down_win = bearish_bbr_df.query("Max_Down > Max_Up.abs()")["Date"].count()
max_down_loss = bearish_bbr_df.query("Max_Down <= Max_Up.abs()")["Date"].count()
total_max_down_trades = max_down_win + max_down_loss
max_down_win_rate = round((max_down_win / total_max_down_trades) * 100 ,2)

max_down_pips = bearish_bbr_df["Max_Down"].sum().round(6)
max_up_pips = bearish_bbr_df["Max_Up"].sum().round(6)
total_max_down_gain = max_down_pips + max_up_pips

average_max_up = bearish_bbr_df["Max_Up"].mean().round(6)
average_max_down = bearish_bbr_df["Max_Down"].mean().round(6)


print(
    f"""
    ========= Initial Bear_BBR Statistics =============

    Total Bear_BBR Signals: {total_signals_count}

    ** Where the move down from the signal resulted **
    ** in a positive gain at the close of the day **
    Day Closed Lower (Count): {day_closed_lower_count}
    Day Closed Higher (Count): {day_closed_higher_count}
    Day Closed Lower (%): {day_closed_lower_pct}
    Day Closed Lower (Pips): {day_closed_lower_pips}

    ** Where after the signal the max move down on the day **
    ** was greater than the max move up **
    D_Max_Down > D_Max_Up (Count): {d_max_down_greater_count}
    D_Max_Down > D_Max_Up (%): {d_max_down_greater_pct}

    ** Where the max move down after the signal **
    ** was greater than the max move up (exit if price closes above signal high) **
    Max_Down > Max_Up (Count): {max_down_win}
    Max_Down <= Max_Up (Count): {max_down_loss}
    Total Trades (Max_Down): {total_max_down_trades}
    Max_Down > Max_Up (Win Rate): {max_down_win_rate}

    Max_Down Pips: {max_down_pips}
    Max_Up Pips: {max_up_pips}
    Total Max_Down Gain: {total_max_down_gain}
    
    Average Max_Up Pips: {average_max_up}
    Average Max_Down Pips: {average_max_down}
 
    ==============================================================================
    """
)


    ========= Initial Bear_BBR Statistics =============

    Total Bear_BBR Signals: 359

    ** Where the move down from the signal resulted **
    ** in a positive gain at the close of the day **
    Day Closed Lower (Count): 172
    Day Closed Higher (Count): 187
    Day Closed Lower (%): 47.91
    Day Closed Lower (Pips): -0.028125

    ** Where after the signal the max move down on the day **
    ** was greater than the max move up **
    D_Max_Down > D_Max_Up (Count): 171
    D_Max_Down > D_Max_Up (%): 47.63

    ** Where the max move down after the signal **
    ** was greater than the max move up (exit if price closes above signal high) **
    Max_Down > Max_Up (Count): 166
    Max_Down <= Max_Up (Count): 190
    Total Trades (Max_Down): 356
    Max_Down > Max_Up (Win Rate): 46.63

    Max_Down Pips: 0.728515
    Max_Up Pips: -0.36732
    Total Max_Down Gain: 0.36119500000000004
    
    Average Max_Up Pips: -0.001032
    Average Max_Down Pips: 0.002046
 
    


### Summary

- From the initial numbers, we can see that the Bear_BBR signal occurred `359` times.
- After the signal occuring, the *max price move down* was greater than the *max price move up* on the day, `47.63%` of the time.
- When the signal occurred, `46.63%` of the time the price dropped more than it gained after the signal, before closing above the signal high.
- The total gains where the price fell more than it rose after the signal (before closing above the signal high) was `3611` pips.
- The average number of pips where price moved against the signal was `10.32` pips
- The average number of pips gained before price closed above the signal high was `20.46` pips.

#### What does this tell us?

- We can see just under 50% of the time, the price moves lower after the signal, *before or without closing above the signal high*.
- Additionally, the gains from when the price moves in favour of the signal is greater than when the price moves against the signal. In short the signal generates more gains than losses overall.
- On average, the price moves down nearly double the amount it moves against the signal.

#### What information is missing?

- Although the signal gains more pips overall, this doesn't tell us the best place to take profit to capture the max move down.
- It is not clear what factors make a move down more likely, or when the signal is more likely to fail.


### SMA16 (Signal Candle High as Stop_Price) - Stats

    1. The stats are calculated by filtering all occurrences
    where the Bear_BBR signal is True. 
    
    2. The dataframe is filtered again to only consider 
    occurrences where the Entry_Price was higher than the SMA16.

    3. (Trade Conditions): The stop price is defined as the High of the signal bar.
        3.1. If the stop price is hit before reaching the SMA16, this is
            a `failure`. 
        3.2. If the price retraces up and does not hit the
            stop price, but hits the SMA16 at a higher price than the
            entry price, this is also a `failure`.
        3.3. If the SMA16 is hit and has a positive To_SMA16 value
            before the stop price is hit, this is a `win`.

In [534]:
pd.options.display.max_rows = 50
bearish_bbr_df.round(6) # round all data to 6 decimals

# get all valid trades
valid_trades = bearish_bbr_df.query("Entry_Price > SMA16")
total_valid_signals = valid_trades["Date"].value_counts().sum()

# get dataframe of occurrences where price fell sma16
# before stop price (the high of the entry_price candle) is hit
to_sma16_win = valid_trades.query("To_SMA16 > 0 and Stop_Pips == 0")

# get dataframe of occurrences where price did not reach sma16
# before stop price is hit
to_sma16_fail = valid_trades.query("To_SMA16 == 0")

# get dataframe of occurrences where price reached sma16
# but sma16 was higher than entry price when hit
to_sma16_fail_above_price = valid_trades.query("To_SMA16 < 0")

# count the total pips gain when sma16 target is reached
total_sma16_pip_gain = to_sma16_win["To_SMA16"].sum().round(6)
# count the total pips lost when stop hit before sma16 target
total_sma16_pip_loss = to_sma16_fail["Stop_Pips"].sum().round(6) + to_sma16_fail_above_price["To_SMA16"].sum().round(6)
# calculate the overall pips gained
total_to_sma16_pips = (total_sma16_pip_loss + total_sma16_pip_gain).round(6)

total_sma16_win_count = to_sma16_win[to_sma16_win.columns[0]].count()
total_sma16_fail_count = to_sma16_fail[to_sma16_fail.columns[0]].count() + to_sma16_fail_above_price[to_sma16_fail_above_price.columns[0]].count()

total_sma16_trades = total_sma16_win_count + total_sma16_fail_count
to_sma16_win_rate = round(total_sma16_win_count / total_sma16_trades * 100, 2)

print(
    f"""
    ========= To_SMA16 (Signal Candle High as Stop_Price) Statistics =============

    Total Bear_BBR Signals: {total_signals_count}
    Total Signals Where Entry > SMA16: {total_valid_signals}
    Total SMA16 Trades (Signal High as Stop): {total_sma16_trades}
    
    Total SMA16 Win: {total_sma16_win_count}
    Total SMA16 Loss: {total_sma16_fail_count}
    SMA16 Win Rate: {to_sma16_win_rate}

    Total SMA16 Win Pips: {total_sma16_pip_gain}
    Total SMA16 Loss Pips: {total_sma16_pip_loss}
    Total SMA16 Pips: {total_to_sma16_pips}
 
    ==============================================================================
    """
)


    ========= To_SMA16 (Signal Candle High as Stop_Price) Statistics =============

    Total Bear_BBR Signals: 359
    Total Signals Where Entry > SMA16: 338
    Total SMA16 Trades (Signal High as Stop): 190
    
    Total SMA16 Win: 56
    Total SMA16 Loss: 134
    SMA16 Win Rate: 29.47

    Total SMA16 Win Pips: 0.0496
    Total SMA16 Loss Pips: -0.087126
    Total SMA16 Pips: -0.037526
 
    


#### SMA16 (Signal Candle High as Stop) - Stats Results

The results show using the `High` of the signal candle as the Stop_Price
results in a negative expectancy.

Out of 359 signals, 338 were selected as meeting the condition where
the entry price was higher than the SMA16. This is done because it 
doesn't make sense using the SMA16 as a target when it is higher than
the entry price. 

Although the total signals were 338, the orginal dataframe gets the
time windows from the signal entry until the price closes above the
signal candle high. 

For this evaluation where the Stop_Price is the
High of the signal bar, we only want to consider a winning signal
as a trade that did not touch or go above the High of the signal bar.
To calaculate this, we take the count of signals where Stop_Pips is 0.
To calculate the failures, we take the count of signals where Stop_Pips
was less than 0, and cases where SMA16 was less than 0.

This only selects 190 trades which amounts to `53%` of 
the overall signals, with only `29.47%` being winning trades. 

Overall this approach would have lost 375 pips.


### SMA16 (Exit if Close > Signal High) - Stats

    1. The stats are calculated by filtering all occurrences
    where the Bear_BBR signal is True. 
    
    2. The dataframe is filtered again to only consider 
    occurrences where the Entry_Price was higher than the SMA16.

    3. (Trade Conditions): The exit condition is defined as the High of the signal bar.
        3.1. If the price closes above the signal high before reaching 
            the SMA16, this is a `failure`. 
        3.2. If the price retraces up and does not meet the exit
            condition, but hits the SMA16 at a higher price than the
            entry price, this is also a `failure`.
        3.3. If the SMA16 is hit and has a positive To_SMA16 value
            before the exit condition, this is a `win`.

In [821]:
pd.options.display.max_rows = 100
bearish_bbr_df.round(6) # round all data to 6 decimals

# get all valid trades
valid_trades = bearish_bbr_df.query("Entry_Price > SMA16")
total_valid_signals = valid_trades["Date"].value_counts().sum()

# get dataframe of occurrences where price fell sma16
# before exit condition (close above the high of the entry_price candle)
to_sma16_win = valid_trades.query("To_SMA16 > 0")

# get dataframe of occurrences where price did not reach sma16
# before exit condition
to_sma16_fail = valid_trades.query("To_SMA16 == 0")

# get dataframe of occurrences where price reached sma16
# but sma16 was higher than entry price when hit
to_sma16_fail_above_price = valid_trades.query("To_SMA16 < 0")

# count the total pips gain when sma16 target is reached
total_sma16_pip_gain = to_sma16_win["To_SMA16"].sum().round(6)
# count the total pips lost when exit condition occurs before sma16 target
total_sma16_pip_loss = to_sma16_fail["Max_Up"].sum().round(6) + to_sma16_fail_above_price["To_SMA16"].sum().round(6)
# calculate the overall pips gained
total_to_sma16_pips = (total_sma16_pip_loss + total_sma16_pip_gain).round(6)

total_sma16_win_count = to_sma16_win[to_sma16_win.columns[0]].count()
total_sma16_fail_count = to_sma16_fail[to_sma16_fail.columns[0]].count() + to_sma16_fail_above_price[to_sma16_fail_above_price.columns[0]].count()

total_sma16_trades = total_sma16_win_count + total_sma16_fail_count
to_sma16_win_rate = round(total_sma16_win_count / total_sma16_trades * 100, 2)

print(
    f"""
    ========= To_SMA16 (Exit if Close > Signal High) Statistics =============

    Total Bear_BBR Signals: {total_signals_count}
    Total Signals Where Entry > SMA16: {total_valid_signals}
    Total SMA16 Trades (Signal High as Stop): {total_sma16_trades}
    
    Total SMA16 Win: {total_sma16_win_count}
    Total SMA16 Loss: {total_sma16_fail_count}
    SMA16 Win Rate: {to_sma16_win_rate}

    Total SMA16 Win Pips: {total_sma16_pip_gain}
    Total SMA16 Loss Pips: {total_sma16_pip_loss}
    Total SMA16 Pips: {total_to_sma16_pips}
 
    ==============================================================================
    """
)


    ========= To_SMA16 (Exit if Close > Signal High) Statistics =============

    Total Bear_BBR Signals: 359
    Total Signals Where Entry > SMA16: 338
    Total SMA16 Trades (Signal High as Stop): 338
    
    Total SMA16 Win: 204
    Total SMA16 Loss: 134
    SMA16 Win Rate: 60.36

    Total SMA16 Win Pips: 0.137636
    Total SMA16 Loss Pips: -0.14225100000000002
    Total SMA16 Pips: -0.004615
 
    


#### SMA16 (Exit if Close > Signal High) - Stats Results

When using an exit condition such as 'waiting for a close above the signal high' instead of 
the Stop_Price, there is a higher win rate of `60.36%`. 

Also a larger number of trades (338) are executed out of the total Bear_BBR signals.

This is an improvement, but there is still a negative expectency.
The approach loses 46 pips overall.

## Investigating factors that increase signal accuracy

### Holding a short position from signal entry to close of the day



In [835]:
def daily_pip_gain(df: Series, pct_adr: int):
    d_pip_gain = df["D_Gain"] if abs(df["D_Max_Up"]) < (df["ADR"] * pct_adr) else (df["ADR"] * -pct_adr)
    return d_pip_gain

def condition_stats(df: DataFrame, pct_adr: int):

    df["Pip_Gain"] = df.apply(daily_pip_gain, axis=1, args=[pct_adr])
    df["ADR%"] = df.apply(lambda x: x["ADR"] * pct_adr, axis=1)
    
    success_series = df.query(f"D_Max_Up.abs() < (ADR * {pct_adr})")["Pip_Gain"]
    fail_series = df.query(f"D_Max_Up.abs() >= (ADR * {pct_adr})")["Pip_Gain"]

    success_total = success_series.count()
    fail_total = fail_series.count()
    total_trades = success_total + fail_total
    
    win_rate = round(success_total / total_trades * 100,2)
    condition_total_pips = round(success_series.sum() + fail_series.sum(),6)

    return {
        "Pct_ADR": pct_adr,
        "Pct_ADR_Avg": df["ADR%"].mean().round(6),
        "ADR_Avg": df["ADR"].mean().round(6),
        "Total Trades": total_trades,
        "Win": success_total,
        "Loss": fail_total,
        "Win Rate": win_rate,
        "Avg_win": success_series.mean().round(6),
        "Avg_Loss": fail_series.mean().round(6),
        "Total Pips": condition_total_pips
    }
    
# columns
cols1 = [
    "Date", "Entry_Price", "Pip_Gain", "D_Max_Up", "ADR%", "D_Gain", "ADR"]
#df
d_gain_df = bearish_bbr_df
# conditions
slope_lt_45_df = d_gain_df.query("SMA32_Slope < 45")


### Using a stop loss set to a percentage of the ADR

Since we already know from the intial stats that the price will close lower on the day 48% of the time after the signal, we should try to limit the losses for the 52% of the times where it fails.

A `stop loss` is used to limit losses, and can be static (e.g. 10 pips) or dynamic (e.g. set as multiple of the `ATR`, or a trailing stop ... etc.).
The ATR takes into account the volatility of the last *x* candles, so is usually a good choice. 

For this particular scenario however, when the `Bear_BBR` signal occurs we are looking to hold a short position until the close of the day. The ATR is not the best approach here, because ATR will look at the average candle range within increments of the timeframe period (e.g. 15min) over *x* period of candles. 

To address this I will use a percentage of the the average daily range (**ADR**) as a stop loss. The ADR calculates the average daily price movement over last 30 days of trading, which will tell me the average number of pips the price moves within a day over that period.

To find a reasonable stop loss range, I'll iterate through different ADR percentages from 1 to 100.

In [837]:
d_gain_experimental_sl = [condition_stats(d_gain_df, x * 0.01) for x in range(1,101)]
d_gain_ex_sl_res_df = pd.DataFrame(d_gain_experimental_sl)
d_gain_ex_sl_res_df

,Pct_ADR,Pct_ADR_Avg,ADR_Avg,Total Trades,Win,Loss,Win Rate,Avg_win,Avg_Loss,Total Pips
0,0.01,0.000091,0.00908,319,10,309,3.13,0.003507,-0.000091,0.007012
1,0.02,0.000182,0.00908,319,12,307,3.76,0.003565,-0.000182,-0.012991
2,0.03,0.000272,0.00908,319,23,296,7.21,0.003789,-0.000272,0.006551
3,0.04,0.000363,0.00908,319,26,293,8.15,0.003751,-0.000363,-0.008811
4,0.05,0.000454,0.00908,319,31,288,9.72,0.003717,-0.000454,-0.015567
5,0.06,0.000545,0.00908,319,39,280,12.23,0.003917,-0.000545,0.000092
6,0.07,0.000636,0.00908,319,47,272,14.73,0.003584,-0.000635,-0.004154
7,0.08,0.000726,0.00908,319,51,268,15.99,0.003519,-0.000725,-0.014929
8,0.09,0.000817,0.00908,319,54,265,16.93,0.003467,-0.000816,-0.029098
9,0.10,0.000908,0.00908,319,59,260,18.50,0.003707,-0.000907,-0.016967


- This shows the stop_loss is directly related to the win rate.
- It also shows that even with a higher win rate, it does not mean a higher `total pips` value. This makes sense because the larger the stop loss is, the larger the losses will be.
- The ideal stop loss will be as low as possible and have a reasonably high `total pips`.
- The maximum total pips returned was 0.073820 which was a result of 28% ADR stop loss, however we can't just take this at face value.